### ISSUE TITLE
 Marionette:Quit is sent twice during session deletion

### ISSUE THREAD

**Issue by Debanjan-B** [mozilla/geckodriver#1235]
## System

* Version: GeckoDriver 64 bit v0.20.0
* Platform: Windows8, Pro 64 bit
* Firefox: 59.0.1 (64-bit)
* Selenium: Selenium-Java v3.11.0

## Testcase : **`webdriver::server`** tries to delete a session twice. As the first call to the DELETE endpoint by calling the **DELETE /session/00db9314-95bf-4642-9eb1-99239b5fad22** is successful the consecutive call **37:[0,5,"quit",{"flags":["eForceQuit"]}]** causes an error.

Perhaps, I am not sure if there is anything different in **4** and **5** as in : 

    37:[0,4,"quit",{"flags":["eForceQuit"]}]
and 

    37:[0,5,"quit",{"flags":["eForceQuit"]}] 

The second `DELETE` call : 
```
1522320308568	Marionette	TRACE	0 -> [0,5,"quit",{"flags":["eForceQuit"]}]
1522320308570	Marionette	TRACE	0 <- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInva ... et@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308571	geckodriver::marionette	TRACE	<- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInvalidSessionIDError@chrome://marionette/content/error.js:327:5\nassert.that/<@chrome://marionette/content/assert.js:407:13\nassert.session@chrome://marionette/content/assert.js:81:3\ndespatch@chrome://marionette/content/server.js:554:7\nexecute@chrome://marionette/content/server.js:531:11\nonPacket/<@chrome://marionette/content/server.js:506:15\nonPacket@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308590	Marionette	DEBUG	Closed connection 0
```

The error : 
```
1522320308568	Marionette	TRACE	0 -> [0,5,"quit",{"flags":["eForceQuit"]}]
1522320308570	Marionette	TRACE	0 <- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInva ... et@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308571	geckodriver::marionette	TRACE	<- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInvalidSessionIDError@chrome://marionette/content/error.js:327:5\nassert.that/<@chrome://marionette/content/assert.js:407:13\nassert.session@chrome://marionette/content/assert.js:81:3\ndespatch@chrome://marionette/content/server.js:554:7\nexecute@chrome://marionette/content/server.js:531:11\nonPacket/<@chrome://marionette/content/server.js:506:15\nonPacket@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308590	Marionette	DEBUG	Closed connection 0
```

Which eventually causes : 

```
###!!! [Child][MessageChannel::SendAndWait] Error: Channel error: cannot send/recv

1522320309975	geckodriver::marionette	ERROR	Failed to stop browser process
```

Minimum Complete Verifiable Example  : 

```
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxDriverLogLevel;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.remote.DesiredCapabilities;

public class A_Firefox 
{
	public static void main(String[] args) 
	{
		System.setProperty("webdriver.gecko.driver", "C:/Utility/BrowserDrivers/geckodriver.exe");
		FirefoxOptions options = new FirefoxOptions();
		options.setLogLevel(FirefoxDriverLogLevel.TRACE);
		WebDriver driver =  new FirefoxDriver(options);
		driver.get("https://stackoverflow.com");
		System.out.println("Application opened");
		System.out.println("Page Title is : "+driver.getTitle());
		driver.quit();
	}
}
```

## Stacktrace : 

First attempt : 

```
1522320308428	webdriver::server	DEBUG	-> DELETE /session/00db9314-95bf-4642-9eb1-99239b5fad22 
1522320308428	geckodriver::marionette	TRACE	-> 37:[0,4,"quit",{"flags":["eForceQuit"]}]
1522320308430	Marionette	TRACE	0 -> [0,4,"quit",{"flags":["eForceQuit"]}]
1522320308432	Marionette	DEBUG	New connections will no longer be accepted
1522320308536	Marionette	TRACE	0 <- [1,4,null,{"cause":"shutdown"}]
1522320308560	geckodriver::marionette	TRACE	<- [1,4,null,{"cause":"shutdown"}]
```

Second attempt : 

```
1522320308560	webdriver::server	DEBUG	Deleting session
1522320308561	geckodriver::marionette	TRACE	-> 37:[0,5,"quit",{"flags":["eForceQuit"]}]
Unable to read VR Path Registry from C:\Users\AtechM_03\AppData\Local\openvr\openvrpaths.vrpath
[Child 5676, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
[Child 5676, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
1522320308568	Marionette	TRACE	0 -> [0,5,"quit",{"flags":["eForceQuit"]}]
1522320308570	Marionette	TRACE	0 <- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInva ... et@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308571	geckodriver::marionette	TRACE	<- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInvalidSessionIDError@chrome://marionette/content/error.js:327:5\nassert.that/<@chrome://marionette/content/assert.js:407:13\nassert.session@chrome://marionette/content/assert.js:81:3\ndespatch@chrome://marionette/content/server.js:554:7\nexecute@chrome://marionette/content/server.js:531:11\nonPacket/<@chrome://marionette/content/server.js:506:15\nonPacket@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308590	Marionette	DEBUG	Closed connection 0
```

## Trace-level log

```
1522320296410	geckodriver	INFO	geckodriver 0.20.0
1522320296422	geckodriver	INFO	Listening on 127.0.0.1:16806
1522320297252	mozrunner::runner	INFO	Running command: "C:\\Program Files\\Mozilla Firefox\\firefox.exe" "-marionette" "-profile" "C:\\Users\\ATECHM~1\\AppData\\Local\\Temp\\rust_mozprofile.CjcldZxrmJVe"
1522320298372	Marionette	DEBUG	Received observer notification "profile-after-change"
1522320298736	Marionette	DEBUG	Received observer notification "command-line-startup"
1522320298736	Marionette	INFO	Enabled via --marionette
1522320299259	geckodriver::marionette	TRACE	  connection attempt 0/600
1522320301352	Marionette	DEBUG	Received observer notification "sessionstore-windows-restored"
1522320301368	geckodriver::marionette	TRACE	  connection attempt 1/600
1522320301648	Marionette	DEBUG	Setting recommended pref browser.newtabpage.enabled to false
1522320301648	Marionette	DEBUG	Setting recommended pref browser.newtabpage.introShown to true
1522320301650	Marionette	DEBUG	Setting recommended pref toolkit.cosmeticAnimations.enabled to false
1522320301650	Marionette	DEBUG	Setting recommended pref datareporting.policy.dataSubmissionPolicyAccepted to false
1522320301650	Marionette	DEBUG	New connections are accepted
1522320301650	Marionette	INFO	Listening on port 31847
1522320302476	geckodriver::marionette	DEBUG	Connected to Marionette on localhost:31847
1522320302990	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:31866
1522320302991	geckodriver::marionette	TRACE	<- {"applicationType":"gecko","marionetteProtocol":3}
1522320302992	geckodriver::marionette	TRACE	-> 163:[0,1,"newSession",{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}]
1522320302992	Marionette	TRACE	0 -> [0,1,"newSession",{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}]
1522320302994	Marionette	WARN	TLS certificate errors will be ignored for this session
1522320303184	Marionette	DEBUG	Register listener.js for window 4294967297
1522320303214	Marionette	TRACE	0 <- [1,1,null,{"sessionId":"00db9314-95bf-4642-9eb1-99239b5fad22","capabilities":{"browserName":"firefox","browserVersion":"59.0. ... AppData\\Local\\Temp\\rust_mozprofile.CjcldZxrmJVe","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1522320303228	geckodriver::marionette	TRACE	<- [1,1,null,{"sessionId":"00db9314-95bf-4642-9eb1-99239b5fad22","capabilities":{"browserName":"firefox","browserVersion":"59.0.1","platformName":"windows_nt","platformVersion":"6.2","pageLoadStrategy":"normal","acceptInsecureCerts":true,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":6856,"moz:profile":"C:\\Users\\AtechM_03\\AppData\\Local\\Temp\\rust_mozprofile.CjcldZxrmJVe","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1522320303228	webdriver::server	DEBUG	<- 200 OK {"value": {"sessionId":"00db9314-95bf-4642-9eb1-99239b5fad22","capabilities":{"acceptInsecureCerts":true,"browserName":"firefox","browserVersion":"59.0.1","moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":6856,"moz:profile":"C:\\Users\\AtechM_03\\AppData\\Local\\Temp\\rust_mozprofile.CjcldZxrmJVe","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"windows_nt","platformVersion":"6.2","rotatable":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000}}}}
Mar 29, 2018 4:15:03 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
1522320303325	webdriver::server	DEBUG	-> POST /session/00db9314-95bf-4642-9eb1-99239b5fad22/url {"url":"https://stackoverflow.com"}
1522320303325	geckodriver::marionette	TRACE	-> 47:[0,2,"get",{"url":"https://stackoverflow.com"}]
1522320303328	Marionette	TRACE	0 -> [0,2,"get",{"url":"https://stackoverflow.com"}]
1522320303334	Marionette	DEBUG	Received DOM event beforeunload for about:blank
1522320304090	Marionette	DEBUG	Received DOM event pagehide for about:blank
1522320305402	Marionette	DEBUG	Received DOM event DOMContentLoaded for https://stackoverflow.com/
1522320308408	Marionette	DEBUG	Received DOM event pageshow for https://stackoverflow.com/
1522320308410	Marionette	TRACE	0 <- [1,2,null,{}]
1522320308411	geckodriver::marionette	TRACE	<- [1,2,null,{}]
1522320308411	webdriver::server	DEBUG	<- 200 OK {"value": {}}
Application opened
1522320308419	webdriver::server	DEBUG	-> GET /session/00db9314-95bf-4642-9eb1-99239b5fad22/title 
1522320308419	geckodriver::marionette	TRACE	-> 19:[0,3,"getTitle",{}]
1522320308422	Marionette	TRACE	0 -> [0,3,"getTitle",{}]
1522320308424	Marionette	TRACE	0 <- [1,3,null,{"value":"Stack Overflow - Where Developers Learn, Share, & Build Careers"}]
1522320308424	geckodriver::marionette	TRACE	<- [1,3,null,{"value":"Stack Overflow - Where Developers Learn, Share, & Build Careers"}]
1522320308424	webdriver::server	DEBUG	<- 200 OK {"value":"Stack Overflow - Where Developers Learn, Share, & Build Careers"}
Page Title is : Stack Overflow - Where Developers Learn, Share, & Build Careers
1522320308428	webdriver::server	DEBUG	-> DELETE /session/00db9314-95bf-4642-9eb1-99239b5fad22 
1522320308428	geckodriver::marionette	TRACE	-> 37:[0,4,"quit",{"flags":["eForceQuit"]}]
1522320308430	Marionette	TRACE	0 -> [0,4,"quit",{"flags":["eForceQuit"]}]
1522320308432	Marionette	DEBUG	New connections will no longer be accepted
1522320308536	Marionette	TRACE	0 <- [1,4,null,{"cause":"shutdown"}]
1522320308560	geckodriver::marionette	TRACE	<- [1,4,null,{"cause":"shutdown"}]
1522320308560	webdriver::server	DEBUG	Deleting session
1522320308561	geckodriver::marionette	TRACE	-> 37:[0,5,"quit",{"flags":["eForceQuit"]}]
Unable to read VR Path Registry from C:\Users\AtechM_03\AppData\Local\openvr\openvrpaths.vrpath
[Child 5676, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
[Child 5676, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
1522320308568	Marionette	TRACE	0 -> [0,5,"quit",{"flags":["eForceQuit"]}]
1522320308570	Marionette	TRACE	0 <- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInva ... et@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308571	geckodriver::marionette	TRACE	<- [1,5,{"error":"invalid session id","message":"","stacktrace":"WebDriverError@chrome://marionette/content/error.js:172:5\nInvalidSessionIDError@chrome://marionette/content/error.js:327:5\nassert.that/<@chrome://marionette/content/assert.js:407:13\nassert.session@chrome://marionette/content/assert.js:81:3\ndespatch@chrome://marionette/content/server.js:554:7\nexecute@chrome://marionette/content/server.js:531:11\nonPacket/<@chrome://marionette/content/server.js:506:15\nonPacket@chrome://marionette/content/server.js:505:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:500:9\n"},null]
1522320308590	Marionette	DEBUG	Closed connection 0
Unable to read VR Path Registry from C:\Users\AtechM_03\AppData\Local\openvr\openvrpaths.vrpath
[Parent 6856, Gecko_IOThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
Unable to read VR Path Registry from C:\Users\AtechM_03\AppData\Local\openvr\openvrpaths.vrpath
[Child 4468, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
[Child 4468, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
Unable to read VR Path Registry from C:\Users\AtechM_03\AppData\Local\openvr\openvrpaths.vrpath
[Child 2052, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
[Child 2052, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
*** UTM:SVC TimerManager:registerTimer called after profile-before-change notification. Ignoring timer registration for id: telemetry_modules_ping
1522320309522	Marionette	DEBUG	Received observer notification "xpcom-shutdown"
1522320309522	Marionette	DEBUG	Resetting recommended pref browser.newtabpage.enabled
1522320309528	Marionette	DEBUG	Resetting recommended pref browser.newtabpage.introShown
1522320309528	Marionette	DEBUG	Resetting recommended pref toolkit.cosmeticAnimations.enabled
1522320309534	Marionette	DEBUG	Resetting recommended pref datareporting.policy.dataSubmissionPolicyAccepted
Unable to read VR Path Registry from C:\Users\AtechM_03\AppData\Local\openvr\openvrpaths.vrpath
[GPU 6968, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346

###!!! [Child][MessageChannel::SendAndWait] Error: Channel error: cannot send/recv

1522320309975	geckodriver::marionette	ERROR	Failed to stop browser process
1522320310378	webdriver::server	DEBUG	<- 200 OK {"value": {}}

```

**Comment by mo** [comment]
For a long time I was using selenium 3.4.0 + geckodriver 0.16.1 and my tests ran just fine. Recently, I upgraded to selenium 3.8.1 + geckodriver 0.19.1 and ran into this bug. While debugging the issue a bit further I found that selenium 3.5.0 + geckodriver 0.16.1 also hits the bug.

Perhaps it could be this selenium commit (done between 3.4 and 3.5) that causes the bug to appear (I have not confirmed this though):
https://github.com/seleniumhq/selenium/commit/0bb8a9b1e5478ff35ea36a24dacc43edca87278e

I've attached full selenium 3.8.1 + geckodriver 0.19.1 logs below (showing the session double delete):
[full-log-selenium-3.8.1-with-geckodriver-0.19.1.txt](https://github.com/mozilla/geckodriver/files/1864824/full-log-selenium-3.8.1-with-geckodriver-0.19.1.txt)
